portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -14,61 +14,97 @@
-    PreparedQuery results = datastore.prepare(query);
-
-    results.asList(FetchOptions.Builder.withDefaults()).forEach(entity -> {
-      comments.add(Comment.fromEntity(entity));
-    });
+    commentStore = new CommentPersistHelper();
+    commentStore.loadComments();
   }
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    Gson gson = new Gson();
+    // The number of comments to send
+    int commentsToSend;
+    CommentPersistHelper.SortMethod sort;
+
+    try {
+      commentsToSend = Integer.parseInt(request.getParameter(NUMBER_COMMENTS_QUERY));
+      // Revert to default if user specified an invalid number
+      if(commentsToSend <= 0) {
+        commentsToSend = DEFAULT_COMMENT_COUNT;
+      }
+    } catch(NullPointerException e) {
+      // Number of comments is not included, so use default
+      commentsToSend = DEFAULT_COMMENT_COUNT;
+    } catch(NumberFormatException e) {
+      // Number of comments is malformed, so complain about it!
+      throw new IllegalArgumentException(request.getParameter(NUMBER_COMMENTS_QUERY) + 
+                " is an invalid number of comments. Aborting GET...");
+    }
+    try {
+      boolean sortAscending = Boolean.parseBoolean(request.getParameter(SORT_ASCENDING_QUERY));
+      if(sortAscending) {
+        sort = CommentPersistHelper.SortMethod.ASCENDING;
+      } else {
+        sort = CommentPersistHelper.SortMethod.DESCENDING;
+      }
+    } catch(NullPointerException e) {
+      // Sort method is not included, so use default
+      sort = DEFAULT_SORT;
+    }
+
     response.setContentType("application/json;");
-    response.getWriter().println(gson.toJson(comments));
+    response.getWriter().println(commentStore.stringifyComments(commentsToSend, sort));
+    // Send the total number of comments
+    response.addIntHeader(TOTAL_NUMBER_HEADER, commentStore.getNumberComments());
   }
 
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
     // Create and add the new comment
-    Comment comment = Comment.fromHttpRequest(request);
-    comments.add(comment);
-
-    // Store the comment so it persists
-    Entity entity = comment.toEntity();
-    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-    datastore.put(entity);
+    commentStore.addComment(request);
     
     // Redirect back to the home page.
     response.sendRedirect("/index.html");
   }
+
+  @Override
+  public void doDelete(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String[] toDelete = request.getParameterValues(DELETE_PARAMETER);
+    Arrays.stream(toDelete).forEach(idStr -> {
+      try {
+        long id = Long.parseLong(idStr);
+        commentStore.deleteComment(id);
+      } catch(NumberFormatException e) {
+        throw new IllegalArgumentException(
```

Comment:
Here is a good resource that we use at Google for defining API best practices. https://google.aip.dev/ *see if you can ignore the syntax for now, unless you're already familiar with google protobufs*. 

One in particular, https://google.aip.dev/235, covers batch delete behaviors and gives this guidance: "The operation should be atomic: it should fail for all resources or succeed for all resources (no partial success)"

I think this would tend to make sense: the ClientApplication has sent an invalid param list, so rather than making it too complicated as to always return which ones failed/not, just prevent any comments from being deleted, until the ClientApplication gives a correct param list.

Reply:
One way to follow this guidance is to separate out the responsibilities: 

1. Iterate over all the Ids and parse them. Upon success, 
2. Delete all the comments.

Reply:
Yup I was wondering about that. Aborting all deletes makes the most sense!

In terms of bulk deletion, it would require using GCloud Dataflow and GQL, which probably lie outside the scope of this project. I can look into it if you like though!

Reply:
You're right, I'd call that out of scope :)